packages/asset-swapper/src/utils/market_operation_utils/balancer_utils.ts

```diff
@@ -42,10 +46,52 @@ export class BalancerPoolsCache {
         return Promise.race([this._getPoolsForPairAsync(takerToken, makerToken), timeout]);
     }
 
-    protected async _getPoolsForPairAsync(takerToken: string, makerToken: string): Promise<BalancerPool[]> {
+    public getCachedPoolAddressesForPair(
+        takerToken: string,
+        makerToken: string,
+        cacheExpiryMs?: number,
+    ): string[] | undefined {
+        const key = JSON.stringify([takerToken, makerToken]);
+        const value = this._cache[key];
+        if (cacheExpiryMs === undefined) {
+            return value === undefined ? [] : value.pools.map(pool => pool.id);
+        }
+        const minTimestamp = Date.now() - cacheExpiryMs;
+        if (value === undefined || value.timestamp < minTimestamp) {
+            return undefined;
+        } else {
+            return value.pools.map(pool => pool.id);
+        }
+    }
+
+    public howToSampleBalancer(
```

Comment:
we will sometimes sample Balancer off-chain, sometimes on-chain. We need to make a GraphQL query once in a while so that the cached pool addresses are kept reasonably fresh, and if we're going to be making the query anyways we might as well compute the quotes off-chain (as we're doing now). Then as long as the cached pool addresses are reasonably fresh, we go with the on-chain sampling approach. 
For the purpose of two-hop, we only use the on-chain sampling approach in order to avoid making two queries (GraphQL + eth RPC) in sequence. As a result it takes a few queries for the Balancer cache to "warm up" and be usable for two-hop